### DISCUSSION TITLE
Turning Plain Old C# Methods to Semantic Kernel Functions

### DISCUSSION THREAD

**Discussion by efficienthacks** [microsoft/semantic-kernel#1694] [General]
I got really excited while reading this: [Adding native functions to the kernel](https://review.learn.microsoft.com/en-us/semantic-kernel/ai-orchestration/native-functions?branch=main&tabs=Csharp) as I have a lot of static methods that do different things that I'd like to chain together.

However, it seems that I'd have to write a new method to get this to be added as an SK plugin:

```cs
[SKFunction("Adds two numbers together")]
[SKFunctionContextParameter(Name = "input", Description = "The first number to add")]
[SKFunctionContextParameter(Name = "number2", Description = "The second number to add")]
public string Add(SKContext context)
{
    return (
        Convert.ToDouble(context["input"]) + Convert.ToDouble(context["number2"])
    ).ToString();
}
```

I'm trying to think of ideas of the easiest way I can get this to work since this is tedious for all the methods I want to add as SK Functions.

I'm not tool well versed in C#, but seems possible to (and would be really nice to have if we could) annotate existing methods with descriptions of the parameters (without converting from string to some type, then back into a string)

```cs
[SKFunction("Adds two numbers together")]
[SomeAttributeToDescribeParam(Name = "first", Description = "The first number to add")]
[SomeAttributeToDescribeParam(Name = "second", Description = "The second number to add")]
public static double Add(double first, double second)
{
    return first + second;
}
```

Please give me any ideas of how to easily convert my existing static c# methods into SKFunctions.  (Preferably without calling a code generator)

Thanks!

**Comment by stephentoub** [comment]
With changes that were merged into SK last week, you can now write your function as almost exactly what you want to write.  It'll now be:
```C#
[SKFunction, Description("Adds two numbers together")]
public static double Add([Description("The first number to add")] double first, [Description("The second number to add")] double second)
{
    return first + second;
}
```
The `[SKFunction]` part is also only needed when importing a whole skill, so that the importer knows which methods to pick.  It's optional if you're just creating an ISKFunction from a MethodInfo or Delegate.  The `[Description("")]` attributes are also optional, but without them you're not providing as much information to the LLM.

I believe @shawncal was working on getting a new version of the nuget published that has this support.  He can comment on whether that's happened yet.

**Reply by efficienthacks** [reply]
This is great news.  Thanks!  Hopefully this will get published soon.  Also hope there is some way to allow the programmer to specify how to convert complex objects to string (e.g.  automatically json-serialize the returned object if it isn't a primitive).  Thanks again!

**Reply by stephentoub** [reply]
> Also hope there is some way to allow the programmer to specify how to convert complex objects to string (e.g. automatically json-serialize the returned object if it isn't a primitive)

In the current implementation it uses TypeConverter.  You can adorn the custom type with [TypeConverter(...)] to specify the converter to be used.

**Comment by efficienthacks** [comment]
I just noticed that this has been pushed in pre-release nuget package: https://github.com/microsoft/semantic-kernel/releases/tag/dotnet-0.17.230626.1-preview

This is the PR for reference:
https://github.com/microsoft/semantic-kernel/pull/1195

Thanks again!
